Adding GEP 726: Path Rewrites and Redirects

robscott opened the pull request:
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
This adds GEP #726, a proposed expansion of redirect capabilities along with new rewrite capabilities. 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

Deploy Preview: https://deploy-preview-731--kubernetes-sigs-gateway-api.netlify.app/geps/gep-726/

robscott commented:
/cc @bowei @hbagdi @howardjohn @jpeach @youngnick

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+          hostname: foo.com
+          path:
+            type: Prefix
+            value: /bar
```
Suggest that we use something like `/app` as the matching prefix. Then we can give a clear example of exactly what redirect location header is generated.

robscott replied:
Updated to use a clearer example for both this and rewrites.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  // the path redirect value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this redirect MUST NOT be applied
+  // to them.
+  PrefixHTTPPathRedirect HTTPPathRedirectType = "Prefix"
```
So, for example,

* the request path is `/api/foo/bar
* and the route is a prefix match on `/api`
* and there is a prefix redirect filter with a path of `/overthere`

The final location path would be `/overthere/foo/bar`?

robscott replied:
Yep, that's exactly the goal, will work on a better explanation for this.

robscott replied:
Updated to use `ReplacePathPrefixMatch`. Given @jpeach's comment below, I think `ReplacePathMatch` could be a simpler option long term, but we'd need to be very careful when it came to documenting support levels (ie some implementations can only support replacing prefix path matches). In any case, we'll also have to document how this is handled if multiple matches and/or match types are present.

robscott replied:
OK, I've been thinking about it more, and I think `ReplacePathMatch` is going to provide significantly more flexibility long term. As noted above, it does mean that we'll need to document the subset of path match types that implementations support, but it seems acceptable to continue to consider 100% of this "Extended" support.

robscott replied:
This has been addressed in more detail in our latest community meeting + https://github.com/kubernetes-sigs/gateway-api/pull/731#issuecomment-964682943 - I ended up going back to `ReplacePathPrefixMatch`.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
```
I'm wondering whether these options are unnecessarily constrained? There's really 2 cases

1. Emit the entire location path
2. Generate the location path by replacing the matched portion

In case (2), the matched portion could be a prefix, or a regex, or implementation-defined.

robscott replied:
On 2 I agree that there's room for other kinds of replacement, but I don't think those are quite as portable. I wanted to start with prefix because that seems to be the most universally supportable concept.

hbagdi replied:
So, we are open to adding an ImplementationSpecific option here in the future?

robscott replied:
I think 2 is actually solved now, and 1 always was. This comment came from a time when the options were "Absolute" and "Prefix" which did not scale with other match types. Given our removal of ImplementationSpecific types elsewhere, I don't think it would make sense to add it back without a clear rationale.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  // Support: Extended
+  //
+  // +optional
+  // +kubebuilder:validation:MaxLength=255
```
MinLength=3 (for "h.c" as the minimum hostname)

howardjohn replied:
Hostname should probably be valid without `.` I think? Like `kubernetes` (when in the default namespace) is valid.

howardjohn replied:
Its probably a bad idea in many cases, but not sure it should be rejected?

youngnick replied:
I think it's important to allow single-label rewrites, sadly. People use them as an escape valve all the time.

robscott commented:
Missed adding a hold on this earlier for further discussion. Will remove if/when it appears we have some consensus.

/hold

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  //
+  // +optional
+  // +kubebuilder:validation:MaxLength=255
+  Hostname *string `json:"hostname,omitempty"`
```
So another approach is to specify that the header filter can set the ':authority` [pseudo header](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3), and the controller can program the underlying proxy to do the right thing?

Perhaps mention this in alternatives and reject it on the basis that this is more direct and explicit?

bowei replied:
If this is to be extended to do more than just an absolute substitution, what would that look like?

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  // the path rewrite value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this rewrite MUST NOT be applied
+  // to them.
+  PrefixHTTPPathRewrite HTTPPathRewriteType = "Prefix"
```
Same comment as for redirect, can we generalize this?

robscott replied:
To clarify, are you suggesting using the same type for both rewrite and redirect?

robscott replied:
I think my latest commit has resolved this, let me know if not.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  // Value defines the HTTP path value to rewrite.
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Value HTTPPathRewriteType `json:"value"`
```
Given this structure, how would we add a regex rewrite type? I think that we would need to add a whole new filter type. That's OK, but wanted to confirm your thinking in this area.

robscott replied:
Yeah I should spell that out. I was actually thinking of adding an optional `regex` or `pattern` field here that was only valid with certain rewrite types.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,266 @@
+  // Type defines the type of path rewrite.
+  //
+  // +kubebuilder:validation:Enum=Prefix
+  Type HTTPPathRewriteType `json:"type"`
```
Does `Absolute` make sense here too? ie. rewrite `/app/foo/any/thing` to `/new/path`?

robscott replied:
I think it would, it just wasn't quite as broadly supported. I guess any implementation that supports regex would also be able to accomplish this though.

robscott replied:
Added absolute as an option in both places as part of transition to shared types.

kmai commented:
I believe this proposal can add a lot of value to enrich routing capabilities in Kubernetes.

Most gateway implementations don't have a normalized API and that makes interchangeability very complicated.

I think it's the best way to encourage modularity in api management tools, edge gateways and service meshes.

youngnick reviewed (commented):
I hadn't taken a look at this yet, this is looking good on an initial read to me. I would guess it's coming post v1alpha2 though.

robscott commented:
Hey @dm3ch, thanks for checking in on this! We actually just discussed it at our community meeting today. It missed the cut for the initial v1alpha2 release (v0.4.0), but now that that's out, this is back in focus. I'm hoping to get this in for the next release of the API.

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,270 @@
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
```
```suggestion
spec:
  hostnames:
  - example.com
```

robscott replied:
Updated, thanks!

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,270 @@
+const (
+  // This type of rewrite indicates that the prefix match will be replaced by
+  // the path rewrite value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this rewrite MUST NOT be applied
```
What if I want to route based on hostname and not path, but I still want to strip a part of the path?
So example.com/foo/bar/baz goes to internal.example.com/bar/baz? This is a very common use-case where users use the routing infrastructure to "fix" such things.

robscott replied:
I think this would still be possible though? The prefix match would be "foo" and the redirect/rewrite value would be "". I need to document that that's possible, but I don't think there's currently anything that would prevent that..

robscott replied:
Added some clarification here, may have overcomplicated the wording though.

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,274 @@
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
```
Why not core?

My concern is that field-level support-level within filters is going to result in end-user UX issues.
This patch would not introduce the issue (we already have Scheme that is extended) but it is increasing the scope of the problem a bit.

robscott replied:
This is tricky because not all types of path redirects are broadly supported. I think it's likely safe to say that absolute and prefix redirects are broadly supportable, not so clear for regex.

hbagdi commented on site-src/geps/gep-726.md:
````diff
@@ -0,0 +1,274 @@
+```go
+// HTTPRequestRewrite defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
````
So, we essentially have two ways to munge a request depending on if it is being forwarded or if it is being redirected. Did you explore the alternative of one filter to munge the request and if it is a redirect, then this another filter can be used?
Is this something worth mentioning in the alternatives considered section for posterity?

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,274 @@
+// HTTPRequestRewrite defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
+type HTTPRequestRewrite struct {
```
What is the conformance level of this filter?

robscott replied:
Apparently I defined conformance at a field level instead of for the filter as a whole, but since both fields have extended conformance, I should probably move that up a level.

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,274 @@
+}
+
+// HTTPPathRewriteType defines the type of path rewrite.
+type HTTPPathRewriteType string
```
I think having a single path rewrite type would the API more intuitive. Is it possible to combine this and HTTPPathRedirectType types and use the same type in both filters?
I think the cognitive load on the user is going to be high otherwise.

robscott replied:
To clarify do you mean that Rewrite + Redirect should not overlap in functionality and instead should be used together when both sets of capabilities are needed?

hbagdi replied:
No. I mean that HTTPPathRedirectType and HTTPPathRewriteType types should be merged into one.

robscott commented:
@bowei @hbagdi @howardjohn @Jeff-Apple @jpeach @youngnick (and anyone else I may have missed), I've updated this PR based on our discussion yesterday at the community meeting. The end result is something a bit closer to what I started with, most notably moving back from "ReplacePathMatch" to "ReplacePrefixPathMatch". 

I've added "ReplacePathMatch" to the alternatives section of the GEP with a set of reasons for why I don't think it is as helpful as it may have seemed on the surface.

I've also added a "future extension" section that describes how we can add additional path rewrite/redirect types in the future. Importantly this shows how regex substitution could work, and I think it ends up being preferable to the more generic "ReplacePathMatch" approach.

Interested in what you think of these changes, hopefully the updated direction makes sense.

hbagdi commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,336 @@
+  //
+  // +optional
+  Path *HTTPPathRewrite `json:"path,omitempty"`
+}
```
There will be two filters: `RequestHeaderModifier` and a `RequestRewrite`, the UX seems a bit off here.
Options are:
- Deprecated RequestHeaderModifier and fold it into this new filter
- Introduce more targetted filters (break this new filter into two): RequestHostnameModifier, RequestPathModifier
- Accept this emergent behavior
- Something else?

robscott replied:
Although I agree the filters are similar in purpose, I think "Rewrite" is a common and reasonably well understood concept here. In most of the implementations I can find, rewriting is limited to either the request hostname and path or the full URL. There are a small set of implementations that extend "rewrite" terminology to headers, but that does not seem to be nearly as consistent.

I think the term "rewrite" is important here given its widespread use throughout other implementations. I think this is going to be a common term that users are looking for. Although it may technically more consistent to have a `RequestModifier` filter or separate `RequestHostnameModifier` and `RequestPathModifier` filters, I think it would lead to more confusion.

We could also bundle header modification into this proposed rewrite filter, but I'm not sure that provides sufficient value to introduce a breaking change and/or duplicate capabilities.

hbagdi replied:
So this is where implementations doing a thing shouldn't be governing our design, right?
The API is to express intent, how the implementations do shouldn't matter - as long as implementations can satisfy the intent.
With that bit, from a purely product and API perspective, would a user not expect a RequestRewrite filter to also be able to modify headers because headers are part of the request?

As a user, wouldn't you be surprised if you had a filter names "RewriteRequest" which won't be able to modify headers? And to modify headers, you had to use a filter called "RequetHeaderModifier"?

Being consistent can help us keep complexity under control in this area.
What are your concerns with duplication of header modification in the RequestRewrite filter as well? We could keep the behavior and intent exactly the same, that should minimize overheads.

robscott replied:
>The API is to express intent, how the implementations do shouldn't matter - as long as implementations can satisfy the intent.

💯 Agree with this.

>What are your concerns with duplication of header modification in the RequestRewrite filter as well? We could keep the behavior and intent exactly the same, that should minimize overheads.

Any time we provide multiple ways to do something, we have to handle conflicts between those. We also have to take extra effort to document and explain the difference between the ways to do things.

>As a user, wouldn't you be surprised if you had a filter names "RewriteRequest" which won't be able to modify headers? And to modify headers, you had to use a filter called "RequetHeaderModifier"?

It's hard for me to say since I'm already so far into the API design. Some thoughts:

1. Header modification may happen at any point, including before a redirect.
1. The current proposal considers rewriting distinct from redirection, merging these together would lead to more confusion as capabilities diverged.
1. We could consider renaming "RequestRewrite" to "URLRewrite" which would more clearly communicate the proposed scope. That would be somewhat inconsistent with the existing "RequestRedirect" filter name, but this is also smaller in scope.
1. We could consider moving everything into smaller filters - PathModifier, HostnameModifier, HeaderModifier, SchemeModifier, etc and then have one final filter that indicated all of the above should result in a redirect and not just a rewrite. I'm concerned that this could result in unnecessary confusion around conformance levels + breaking API changes.

robscott replied:
Chatted with @hbagdi a bit and decided that 3 was the most reasonable approach going forward, I've updated the GEP to reflect that. I've also created https://github.com/kubernetes-sigs/gateway-api/issues/928 as a related follow up from our conversation.

youngnick commented on site-src/geps/gep-726.md:
```diff
@@ -238,9 +246,63 @@ The NGINX rewrite module contains the following relevant capabilities
+
+Both of the following can be represented by adding a new field new types. For
+example, this config would result in a request to `/foo/baz` to be rewritten to
+`/foo/bar`:
```
```suggestion
`/bar/baz`:
```
I think?

robscott replied:
Great catch, thanks!

youngnick reviewed (commented):
LGTM aside from one small thing.

hbagdi commented:
/lgtm

ccfishk commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,319 @@
+  requestRewrite:
+    path:
+      type: RegularExpression
+      pattern: /foo/(.*)/baz
```
isn't "*" enough here ?

bowei replied:
I think you are thinking of UNIX glob expression vs regular expression.

The dot "." means any character.

ccfishk commented on site-src/geps/gep-726.md:
````diff
@@ -0,0 +1,319 @@
+```
+
+Similarly, this config would result in a request to `/foo/bar/baz` being
+rewritten to `/foo/other/baz`.
````
Shouldn't be A request to `/foo/baz` being
rewritten to `/bar/baz` ?
Confused.

robscott replied:
This is a purely theoretical future implementation of regex path rewrites, but in this case, users would provide a separate pattern (regex) and substitution. 

```
      pattern: /foo/(.*)/baz
      substitution: other
```

So in this case, the `substitution` would be replacing the `(.*)` part of the pattern.

bowei commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,319 @@
+  // replaced by the substitution value. For example, a path with a prefix match
+  // of "/foo" and a ReplacePrefixPathMatch substitution of "/bar" will have the
+  // "/foo" prefix replaced with "/bar" in matching requests.
+  PathMatchHTTPPathModifier HTTPPathModifierType = "ReplacePrefixPathMatch"
```
I find this naming a bit confusing, "path" appears twice:

*Path*MatchHTTP*Path*Modifier

bowei replied:
Why not just MatchHTTPPathModifier?

jpeach replied:
```suggestion
  ReplacePrefixHTTPPathModifier HTTPPathModifierType = "ReplacePrefix"
```

WDYT?

robscott replied:
This was renamed a few too many times and I missed the actual variable name. Gave this another update, this time to `PrefixMatchHTTPPathModifier` and `ReplacePrefixMatch` respectively. Hopefully that makes a bit more sense.

bowei commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,319 @@
+
+### Redirects
+
+The existing `RequestRedirect` filter can be expanded to support path redirects.
```
To be very clear, redirect means that the server sends a 3xx back to the client?

Which 3xx is sent? Does it matter?

jpeach replied:
the redirect filter sets 302 as the default status code

robscott replied:
Reference: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L703

bowei commented on site-src/geps/gep-726.md:
````diff
@@ -0,0 +1,319 @@
+
+This would be represented with the following API additions:
+```go
+// HTTPRequestRedirect defines a filter that redirects a request. At most one of
````
We should state very explicitly this sends a 3xx response to the client.

robscott replied:
This part of the GEP is confusing because it was not clear that the redirect filter itself already exists and is more clearly defined than this snippet suggests. I've updated the GEP to clarify that this is an addition to an existing filter. https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L703

bowei commented:
Generally LGTM -- probably hash out the details in the actual API review.

jpeach commented on site-src/geps/gep-726.md:
```diff
@@ -0,0 +1,319 @@
+
+  // This type of modifier indicates that any prefix path matches will be
+  // replaced by the substitution value. For example, a path with a prefix match
+  // of "/foo" and a ReplacePrefixPathMatch substitution of "/bar" will have the
```
From an API perspective, I think that I would prefer to decouple prefix matching from prefix rewriting. This is a  bit awkward in Envoy IIRC, but it's doable (at least Contour does it).

If we keep this as it is, then a rewrite would become invalid if the path match changes, and it would also rewrite differently depending on the match prefix. This seems operationally fraught to me, and likely to trick people with unexpected side-effects.

robscott replied:
My understanding of the [Envoy spec](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-redirectaction-prefix-rewrite) is that the only kind of prefix rewriting it supports is based on the corresponding prefix match. That's also true for the `pathPrefixRewrite` in [GCP UrlMaps](https://cloud.google.com/compute/docs/reference/rest/v1/urlMaps#UrlMap.FIELDS.inlinedField_22). Of course with Envoy and other proxies, regex path rewrites could be used to enable more advanced use cases including rewrites based on an arbitrary prefix (covered later in this GEP as possible future extensions), but that doesn't seem to be quite as widely supported as rewriting based on a prefix match.

jpeach replied:
Contour implements this by implicitly expanding it to multiple matches (IIRC I wrote that, but I forget the details). As an API feature, I'd prefer that Envoy limitations didn't leak through, though I'm sympathetic to the complexity of implementing this in the Envoy API.

robscott replied:
>I'd prefer that Envoy limitations didn't leak through

I agree with that, though in this case it's something that also affects GCP APIs, where the proposed `PrefixMatch` is the only path rewrite currently supported. Although I don't want to copy APIs, I also want to build an API that is broadly implementable. I think the current GEP along with the proposed future extensions for Prefix and Regex rewrites based on custom patterns will provide the best possible balance between an implementable and understandable API, but open to other potential approaches.

hbagdi commented:
/lgtm

robscott commented:
Discussed this at today's community meeting, sounded like there was fairly broad consensus, going to remove the hold. Thanks to everyone for all the help with this one!

/hold cancel
